frontend/src/container/FormAlertRules/index.tsx

```diff
@@ -0,0 +1,256 @@
+		});
+	}, [initialValue]);
+
+	const onCancelHandler = useCallback(async () => {
+		history.replace(ROUTES.LIST_ALL_ALERT);
+	}, []);
+
+	const isFormValid = useCallback((): boolean => {
+		let retval = true;
+
+		if (!alertDef.alert || alertDef.alert === '') {
+			notification.error({
+				message: 'Error',
+				description: 'alert name is required',
+			});
+			return false;
+		}
+
+		if (
+			queryCategory === PROMQL &&
+			(!promQueries || Object.keys(promQueries).length === 0)
+		) {
+			notification.error({
+				message: 'Error',
+				description:
+					'promql expression is required when query format is set to PromQL',
+			});
+			return false;
+		}
+
+		if (
+			(queryCategory === QUERY_BUILDER && !metricQueries) ||
+			Object.keys(metricQueries).length === 0
+		) {
+			notification.error({
+				message: 'Error',
+				description: 'at least one metric condition is required',
+			});
+			return false;
+		}
+
+		Object.keys(metricQueries).forEach((key) => {
+			if (metricQueries[key].metricName === '') {
+				retval = false;
+				notification.error({
+					message: 'Error',
+					description: `metric name is missing in ${metricQueries[key].name}`,
+				});
+			}
+		});
+
+		Object.keys(formulaQueries).forEach((key) => {
+			if (formulaQueries[key].expression === '') {
+				retval = false;
+				notification.error({
+					message: 'Error',
+					description: `expression is missing in ${formulaQueries[key].name}`,
+				});
+			}
+		});
+
+		return retval;
+	}, [alertDef, queryCategory, metricQueries, formulaQueries, promQueries]);
+
+	const onSaveHandler = useCallback(async () => {
+		if (!isFormValid()) {
+			return;
+		}
+
+		const postableAlert: AlertDef = {
+			...alertDef,
+			ruleType: queryCategory === PROMQL ? 'promql_rule' : 'threshold_rule',
+			condition: {
+				...alertDef.condition,
+				compositeMetricQuery: {
+					builderQueries: prepareBuilderQueries(metricQueries, formulaQueries),
+					promQueries,
+					queryType: queryCategory,
+				},
+			},
+		};
+		console.log(' postableAlert :', postableAlert);
+
+		setLoading(true);
+		const apiReq =
+			ruleId && ruleId > 0
+				? { data: postableAlert, id: ruleId }
+				: { data: postableAlert };
+		console.log(' ruleId :', ruleId);
+		const response = await saveAlertApi(apiReq);
+
+		if (response.statusCode === 200) {
+			notifications.success({
+				message: 'Success',
+				description:
+					!ruleId || ruleId === 0
+						? 'Rule created successfully'
+						: 'Rule edited successfully',
+			});
+			setTimeout(() => {
+				history.replace(ROUTES.LIST_ALL_ALERT);
+			}, 2000);
+		} else {
+			notifications.error({
+				message: 'Error',
+				description: response.error || 'failed to create or edit rule',
+			});
+		}
+		setLoading(false);
+	}, [
+		isFormValid,
+		queryCategory,
+		ruleId,
+		notifications,
+		alertDef,
+		metricQueries,
+		formulaQueries,
+		promQueries,
+	]);
+
+	const renderBasicInfo = (): JSX.Element => (
```

Comment:
Can we wrap in useMemo

Reply:
facing an issue while memoizing. Will handle in next release